### PR TITLE
Add optional precommit hooks

### PR DESCRIPTION
**Pull request by matthew-d-jones**
Adds optionally installable precommit hooks for `flake8` and `yapf` and documents this in the developer documentation.

### CODE REVIEW

File: docs/developer/getting-started.rst [hunk 1]
```diff
@@ -116,3 +116,13 @@ Building Documentation
...
+
+Precommit Hooks
+---------------
+
+If you wish, you can install precommit hooks for flake8 and yapf. In the source directory run:
+
+.. code-block:: bash
+
+  pre-commit install
+  pre-commit run --all-files
```
Comment (SimonHeybrock) [inline docs/developer/getting-started.rst:128]: Not sure how `pre-commit` works exactly. Our CI setup excludes certain files when running the formatting checks (see azure configs). How is this taken into account here?

Comment (matthew-d-jones) [reply docs/developer/getting-started.rst:128]: I'll check. If exclusions are in the tool's config file then they should be picked up, but that won't work if they are set using cmd line args in the CI scripts.

Comment (SimonHeybrock) [reply docs/developer/getting-started.rst:128]: That would be worth changing then, to be handled in the config files.

Comment (matthew-d-jones) [reply docs/developer/getting-started.rst:128]: I've restricted yapf to only run on files in the "python" directory to match CI.
There is no exclusion on flake8. CI config explicitly globs *.py, but that's redundant; flake8 does it by default.

Comment (SimonHeybrock) [reply docs/developer/getting-started.rst:128]: 👍 Right, the exclusions I remembered were for `clang-format`.

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**